Key Management Guide (#1140)
TBD54566975/developer.tbd.website: site/docs/web5/build/decentralized-identifiers/key-management.mdx

---

> sidebar_position: 2

Reviewer: can you add `sidebar_position: 1` to the 'Create a DID' guide so that the KMS guide appears after that one?

---

# Key Management

> Cryptographic keys are an integral part of DIDs and DIDs are designed to represent secure identity, which is why good key management is essential to developing your Web5 app. Key management provides the optionality to securely store your cryptographic keys and is built right into the `Web5` SDK. 

Reviewer: ```suggestion
[Decentralized Identifiers (DIDs)](/docs/web5/learn/decentralized-identifiers) are associated with public and private keys. The private key, which should be kept secret, is used to sign information or transactions, demonstrating control over the DID. Proper [key management](https://en.wikipedia.org/wiki/Key_management) ensures that private keys are securely stored and used, reducing the risk of unauthorized access or misuse.
```

Reviewer 2: Great rewrite @angiejones

> ## Quickstart

Reviewer: ```suggestion

```

> Key management is built right into the `DidMethod` implementations provided in the `Web5` SDK, such as `DidDht`. Here is how you'll create or load a DID with a `KeyManager`:

Reviewer: ```suggestion
The Web5 SDK provides a key manager by default, as well as the option for you to use your own.

Here's an example that uses Web5's `InMemoryKeyManager` for development environments, and an external [AWS Key Management Service](https://docs.aws.amazon.com/kms/latest/developerguide/overview.html) for production environments:
```

```kotlin
fun initKeyManagement(env: string, didUri: string){
    var keyManager: KeyManager;

    // Determine which key manager you'd like to use based on
    // the environment
    when(env) {
        "development" -> keyManager = InMemoryKeyManager();
        "production" -> keyManager = AwsKeyManager()
    }

    // Initialize or load a DID

>     var did: DidDht;

Reviewer: this has a warning: 'Variable 'did' is never used'

    if (didUri == null) {
        // Create a new DID

>         val did = DidDht.create(keyManager)

Reviewer: this gives a syntax error: 'Variable 'keyManager' must be initialized'

    } else {
        // Load existing DID

>         val did = DidDht(uri = "some_uri", keyManager)

Reviewer: the constructor for DidDht is

```kt
public class DidDht(
  uri: String, 
  keyManager: KeyManager,
  public val didDocument: DIDDocument? = null,
  private val didDhtApi: DidDhtApi
) 
```

    }
}
```

> The `Web5` SDK provides a variety of `KeyManager` implementations, which will allow you connect to storage - in-memory storage in the case of `InMemoryKeyManager` and AWS Key Management Service in the case of `AwsKeyManager` - and allow you to perform all the operations exposed in the `KeyManager` interface. 

Reviewer: I really don't see the need for anything from this paragraph on down, but I'll leave it to @frankhinek and @mistermoe to make the call here

Reviewer 2: There are certainly advanced topics in terms of key management, but I wouldn't recommend organizing them in terms of "database, storage, and management layer."  If we want to include advanced key management, it should likely be written as its own guide that gets into the considerations and customization options for secure key management.  It would likely also need to be divided up by runtime environment since options vary greatly for native mobile, web browser, and backend service.

This is a quick, off the top of my head rewrite:

```suggestion
The Web5 SDK offers two distinct `KeyManager` implementations:

- `InMemoryKeyManager`: This implementation utilizes in-memory, ephemeral storage for managing keys. It's ideal for scenarios where temporary key management is sufficient, such as in testing environments. Since the keys are stored in memory, they are lost when the application is terminated.

- `AwsKeyManager`: This implementation integrates with the AWS Key Management Service (KMS), providing a secure and robust solution for key management. It's well-suited for production environments where a high level of security and availability are critical.

Both implementations facilitate key generation and signature operations essential for creating and managing DIDs. Each caters to distinct security needs, and the choice between them should be based on the specific requirements of your application.
```

This guide will explain the key concepts of key management as they pertain to your app, highlight the design of key management in your app, and give you the concrete code snippets needed to ensure the safety of your keys.

### Database & Storage

Depending on the platform, runtime, and use-case you're developing for, you'll have a variety of different options for key storage. If you're developing a frontend application, you'll likely be able to take advantage of OS or hardware-level features like the Secure Enclave on iPhone or Secure Storage Service on Android. Backend developers may prefer to offload key storage to a service such as AWS' Key Management Service. Lastly, web developers will find it more difficult to store keys within the constraints they have, but have options like the Web Cryptography API available to them.

### Management Layer

Once you've decided on an appropriate storage solution for your cryptographic keys, you'll next need to select how you'd like for those keys to be managed. There are three main types of storage for you to select from on most platforms:

- **In-Memory** storage is best for development and testing environments.

> - **Secure Enclave** storage is normally on mobile devices and other devices with specialized key storage hardware.

Reviewer 3: We cannot take advantage of the Secure Enclave, as it does not support either Ed25519 or Secp256k1. We should be removing any references to Secure Enclave from our guides.

<!-- JS + Kotlin -->
Depending on the platform, runtime, and use-case you're developing for, you'll have a variety of different options for key storage. If you're developing a frontend application, you'll likely be able to take advantage of OS or hardware-level features like the Secure Enclave on iPhone or Secure Storage Service on Android. Backend developers may prefer to offload key storage to a service such as AWS' Key Management Service. Lastly, web developers will find it more difficult to store keys within the constraints they have, but have options like the Web Cryptography API available to them.

### Management Layer

Once you've decided on an appropriate storage solution for your cryptographic keys, you'll next need to select how you'd like for those keys to be managed. There are three main types of storage for you to select from on most platforms:

- **In-Memory** storage is best for development and testing environments.
- **Secure Enclave** storage is normally on mobile devices and other devices with specialized key storage hardware.
- **Remote** storage is best for backend systems where sub-100ms latency is not required and key storage can be abstracted away.

> The `web5/crypto` package contains a variety of key manager implementations, including a basic [`InMemoryKeyManager`]() designed for development environment testing, as well as an [`AwsKeyManager`]() implementation to allow your app to interface with AWS' remote Key Management Service. Each manager is designed for a specific type of storage, and while you are free to develop your own key manager, the `Web5` SDK will be updated with support for new key managers based on feedback. 

Reviewer: build is failing because of the missing links. I assume these are coming?
